In [1]:
import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "../")))

from database_operations import DatabaseOperations
from overtake_model import OvertakingModel
from race_data import RaceDataSetup
from race_dataframe import RaceDataframe
from race_sim import RaceSimulator
from evaluation import RaceSimEvaluation, EvaluateMany

import pandas as pd
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)
pd.set_option('display.max_colwidth', None) 

In [2]:
db1 = DatabaseOperations(2024, "Mexico City")
race1 = RaceDataframe(db1)
race_data1 = RaceDataSetup(db1, race1)
overtake1 = OvertakingModel(race1.race_df)
test = overtake1.get_model_accuracy()[1]

In [3]:
# race1.race_df[race1.race_df["lap_num"]>69]

In [4]:
# print(race_data1.driver_tyre_coefficients)

race_sim = RaceSimulator(race_data1, overtake1)
race_sim.simulate()
print(race_sim.num_overtakes)
sim_df = race_sim.get_results_as_dataframe()
sim_df

84


driver_number       driver_name                 pit_schedule  tyre_type  \
0              55      Carlos Sainz                {33: 3, 1: 2}          3   
1               4      Lando Norris                {31: 3, 1: 2}          3   
2              16   Charles Leclerc         {32: 3, 70: 1, 1: 2}          1   
3              44    Lewis Hamilton                {29: 3, 1: 2}          3   
4              81     Oscar Piastri                {40: 3, 1: 2}          3   
5              63    George Russell                {32: 3, 1: 2}          3   
6               1    Max Verstappen                {27: 3, 1: 2}          3   
7              20   Kevin Magnussen                {31: 3, 1: 2}          3   
8              27   Nico Hulkenberg                {30: 3, 1: 2}          3   
9              10      Pierre Gasly                {29: 3, 1: 2}          3   
10             18      Lance Stroll                {27: 3, 1: 2}          3   
11             45  Franco Colapinto                {48: 2, 1: 3}          2   
12             31      Esteban Ocon                {1: 3, 49: 2}          2   
13             77   Valtteri Bottas                {50: 2, 1: 3}          2   
14             24       Guanyu Zhou                {44: 2, 1: 3}          2   
15             40       Liam Lawson         {40: 2, 66: 1, 1: 3}          1   
16             11      Sergio Perez  {21: 2, 44: 2, 69: 1, 1: 3}          1   
17             23   Alexander Albon                       {1: 2}          2   
18             22      Yuki Tsunoda                       {1: 2}          2   
19             14   Fernando Alonso                       {1: 2}          2   

    lap_num  sector  sector_time  stint_lap  cumulative_time        gap  \
0        71       3    20.659784         39      6263.274114   0.000000   
1        71       3    20.703931         41      6265.449308   2.175194   
2        71       3    22.274851          2      6288.469647  23.020339   
3        71       3    20.863479         43      6299.578047  11.108400   
4        71       3    20.743427         32      6310.674003  11.095956   
5        71       3    20.987247         40      6311.681661   1.007658   
6        71       3    20.817306         45      6317.422278   5.740616   
7        71       3    21.040058         41      6327.308067   9.885789   
8        71       3    20.856500         42      6343.899743  16.591676   
9        71       3    21.015713         43      6363.300474  19.400731   
10       71       3    20.813931         45      6369.164909   5.864434   
11       71       3    21.133117         24      6369.504041   0.339132   
12       71       3    21.052458         23      6371.008173   1.504133   
13       71       3    21.274378         22      6390.646420  19.638247   
14       71       3    21.132451         28      6390.795134   0.148714   
15       71       3    22.196024          6      6392.904418   2.109285   
16       71       3    21.480647          3      6411.543292  18.638873   
17       71       0     0.000000         71         0.000000   0.000000   
18       71       0     0.000000         71         0.000000   0.000000   
19       71       3    23.298221         71      1495.102914   0.131762   

      pit       pace  position  starting_pos  \
0   False  31.549575         1             1   
1   False  20.770952         2             2   
2   False  21.660755         3             3   
3   False  20.932355         4             4   
4   False  20.795339         5             8   
5   False  21.022026         6             5   
6   False  20.890183         7             6   
7   False  21.082077         8             7   
8   False  20.935128         9             9   
9   False  21.082269        10            10   
10  False  20.893399        11            11   
11  False  21.168223        12            12   
12  False  21.090353        13            13   
13  False  21.222762        14            14   
14  False  21.175391        15            15   


In [5]:
eval = RaceSimEvaluation(race_sim, race1, db1)

print(eval.compare_total_cumulative_times()[1])
eval.compare_total_cumulative_times()[0]


272.11586913449133


driver_number       driver_name  simulated_cumulative_time  \
0              55      Carlos Sainz                6263.274114   
1               4      Lando Norris                6265.449308   
2              16   Charles Leclerc                6288.469647   
3              44    Lewis Hamilton                6299.578047   
4              81     Oscar Piastri                6310.674003   
5              63    George Russell                6311.681661   
6               1    Max Verstappen                6317.422278   
7              20   Kevin Magnussen                6327.308067   
8              27   Nico Hulkenberg                6343.899743   
9              10      Pierre Gasly                6363.300474   
10             18      Lance Stroll                6369.164909   
11             45  Franco Colapinto                6369.504041   
12             31      Esteban Ocon                6371.008173   
13             77   Valtteri Bottas                6390.646420   
14             24       Guanyu Zhou                6390.795134   
15             40       Liam Lawson                6392.904418   
16             11      Sergio Perez                6411.543292   
17             23   Alexander Albon                   0.000000   
18             22      Yuki Tsunoda                   0.000000   
19             14   Fernando Alonso                1495.102914   

    actual_cumulative_time  absolute_error  
0                 6023.912      239.362114  
1                 6028.139      237.310308  
2                 6057.475      230.994647  
3                 6067.579      231.999047  
4                 6083.971      226.703003  
5                 6071.151      240.530661  
6                 6083.463      233.959278  
7                 6086.111      241.197067  
8                 6024.711      319.188743  
9                 5959.340      403.960474  
10                6045.905      323.259909  
11                6047.376      322.128041  
12                6053.418      317.590173  
13                6062.091      328.555420  
14                6081.243      309.552134  
15                6085.804      307.100418  
16                6093.613      317.930292  
17                     NaN             NaN  
18                     NaN             NaN  
19                1428.339       66.763914

In [6]:
eval.get_position_accuracy_final_class()

{'position_accuracy': 0.7647058823529411,
 'top_3_accuracy': 1.0,
 'mean_error': 0.35294117647058826,
 'total_error': 6}

In [7]:
eval.compare_total_to_front()

{'total_simulated_gap_to_front': 1310.9637926282703,
 'total_actual_gap_to_front': 1272.1519999999973}

{'total_simulated_gap_to_front': 1305.9637926282421,
 'total_actual_gap_to_front': 1272.1519999999973}

In [8]:
from typing import List, Dict

def find_most_accurate_races(race_inputs: List[Dict]) -> List[Dict]:
	"""
	Finds the 4 most accurately simulated races based on the average mean error.

	Args:
		race_inputs (List[Dict]): A list of dictionaries, where each dictionary contains the inputs
								  for a race simulation (year, circuit, etc.).

	Returns:
		List[Dict]: A list of the 4 most accurate races, sorted by their average mean error (lowest first).
					Each dictionary includes the race inputs and its evaluation metrics.
	"""
	race_evaluations = []

	# Iterate through each race input
	for race_input in race_inputs:
		year = race_input["year"]
		circuit = race_input["circuit"]

		try:
			# Initialize database, race data, setup, overtaking model, and simulation
			db = DatabaseOperations(year, circuit)
			race_data = RaceDataframe(db)
			race_setup = RaceDataSetup(db, race_data)
			overtake = OvertakingModel(race_data.race_df)
			simulation = RaceSimulator(race_setup, overtake)
			simulation.simulate()

			# Evaluate the simulation
			eval = RaceSimEvaluation(simulation, race_data, db)
			metrics = eval.get_position_accuracy_final_class()
			# Store the race inputs and evaluation metrics
			race_evaluations.append({
				"race_input": race_input,
				"metrics": metrics,
				"overtakes": simulation.num_overtakes,
				"has_safety_car": simulation.has_safety_car
			})

		except Exception as e:
			# Log the error and continue with the next race
			print(f"Error evaluating race {year} at {circuit}: {e}")
			continue

	# Sort the races by average mean error (lower is better)
	sorted_races = sorted(
		race_evaluations,
		key=lambda x: x["metrics"]["mean_error"]
	)

	# Return the top 4 most accurate races
	return sorted_races